### PR TITLE
BUG: EWM silently failed float32

### LINKED ISSUE #42452: BUG: pandas EWM fails silently if data types are float32 instead of float64
- [X ] I have checked that this issue has not already been reported.

- [ X] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import numpy as np
import pandas as pd

kk = pd.DataFrame(np.random.rand(20, 3))
kk = kk.astype(np.float32)
print(kk.ewm(alpha=0.5, axis=1).mean().shape)
# (20, 0)

kk = kk.astype(np.float64)
print(kk.ewm(alpha=0.5, axis=1).mean().shape)

# (20, 3)
```

#### Problem description

This should not depend on dtypes.

#### Output of ``pd.show_versions()``

<details>

commit           : 7c48ff4409c622c582c56a5702373f726de08e96
python           : 3.7.10.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.5.0
Version          : Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.5
numpy            : 1.17.4
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 21.1.2
setuptools       : 57.0.0
Cython           : None
pytest           : 6.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.8.5 (dt dec pq3 ext lo64)
jinja2           : 2.10.3
IPython          : 7.23.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : 0.4.1
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : None
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : 0.17.1
pyxlsb           : None
s3fs             : None
scipy            : 1.5.4
sqlalchemy       : 1.3.19
tables           : None
tabulate         : 0.8.9
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by debnathshoham**
- [x] closes #42452
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
cc @mroeschke

File: pandas/tests/window/test_ewm.py [hunk 1]
```diff
@@ -181,3 +181,14 @@ def test_ewma_times_adjust_false_raises():
         Series(range(1)).ewm(
             0.1, adjust=False, times=date_range("2000", freq="D", periods=1)
         )
+
+
+@pytest.mark.parametrize("func", ["mean", "std", "var"])
+@pytest.mark.parametrize("dtype", [np.float32, np.float16, np.float64, float, "float"])
```
Comment (mroeschke) [inline pandas/tests/window/test_ewm.py:187]: Could you use the `float_dtype` pytest fixture instead?

File: pandas/tests/window/test_ewm.py [hunk 2]
```diff
@@ -181,3 +181,14 @@ def test_ewma_times_adjust_false_raises():
...
+def test_float_dtype_ewma(dtype, func):
+    # GH#42452
+    df = DataFrame(np.random.rand(20, 3), dtype=dtype)
```
Comment (mroeschke) [inline pandas/tests/window/test_ewm.py:190]: Could you use constant data, build a `expected = DataFrame(...)` and use `tm.assert_frame_equal(result, expected)`?

**Review by mroeschke — COMMENTED** [review]
Also needs a whatsnew entry for 1.4

File: pandas/tests/window/test_ewm.py (hunk 1 above)
Comment (debnathshoham) [reply pandas/tests/window/test_ewm.py:187]: I don't see a `float_dtype` fixture. Are you suggesting to add that?

Comment (debnathshoham) [reply pandas/tests/window/test_ewm.py]: My bad. Added.

**Review by jreback — CHANGES_REQUESTED** [review]
we certainly don't support float16; this is not actually testing the result is correct

**Comment by debnathshoham** [comment]
Hi @jreback . Sorry I don't follow.
> we certainly don't support float16

Are you saying I should remove float16 here? 
`converted_dtypes.extend([np.float64, np.float32, np.float16])`


>this is not actually testing the result is correct

I have changed the earlier test (where I was just comparing the shape), to include constant data. I am comparing the result now(after @mroeschke suggested).

File: pandas/core/frame.py [hunk 3]
```diff
@@ -4279,6 +4279,11 @@ def check_int_infer_dtype(dtypes):
                     # error: Argument 1 to "append" of "list" has incompatible type
                     # "Type[signedinteger[Any]]"; expected "Type[signedinteger[Any]]"
                     converted_dtypes.append(np.int64)  # type: ignore[arg-type]
+                elif dtype == "float":
+                    # GH#42452 : np.dtype("float") coerces to np.float64 from Numpy 1.20
+                    converted_dtypes.extend(
+                        [np.float64, np.float32, np.float16]  # type: ignore[list-item]
```
Comment (mroeschke) [inline pandas/core/frame.py:4285]: I think here @jreback means we don't explicitly support `np.float64`. We would just like to capture np.float32/64

**Review by mroeschke — CHANGES_REQUESTED** [review]

File: pandas/tests/window/test_ewm.py [hunk 4]
```diff
@@ -181,3 +181,35 @@ def test_ewma_times_adjust_false_raises():
         Series(range(1)).ewm(
             0.1, adjust=False, times=date_range("2000", freq="D", periods=1)
         )
+
+
+@pytest.mark.parametrize("func", ["mean", "std", "var"])
+def test_float_dtype_ewma(func, float_dtype):
+    # GH#42452
+    expected_mean = DataFrame(
```
Comment (mroeschke) [inline pandas/tests/window/test_ewm.py:189]: Could you parameterize the expected results too like:

```
@pytest.mark.parametrize(
    "func, expected_data", 
    [["mean", [...]],
     ["std", [...],
     ...])
```

Testing on 1 column of data is fine as well e.g. `range(5)`

File: pandas/core/frame.py [hunk 5]
```diff
@@ -4279,6 +4279,11 @@ def check_int_infer_dtype(dtypes):
                     # error: Argument 1 to "append" of "list" has incompatible type
                     # "Type[signedinteger[Any]]"; expected "Type[signedinteger[Any]]"
                     converted_dtypes.append(np.int64)  # type: ignore[arg-type]
+                elif dtype == "float":
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/frame.py:4282]: this is really far down the chain from the caller. what exactly is calling this? this has all kinds of implications changing here (which is really puzzling why nothing else breaks)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (debnathshoham) [reply pandas/core/frame.py:4282]: This is being called below in `obj.select_dtypes`

https://github.com/pandas-dev/pandas/blob/b5bceb7271366a0d016a5983a38ddc2b6e86ecf1/pandas/core/window/rolling.py#L218-L232

Yes, ideally the else should have handled this. But this `infer_dtype_from_object` calls `pandas_dtype`, which ultimately calls np.dtype(dtype) as below (which defaults to np.float64)
https://github.com/pandas-dev/pandas/blob/b5bceb7271366a0d016a5983a38ddc2b6e86ecf1/pandas/core/dtypes/common.py#L1775

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/frame.py:4282]: ok can you add some unit tests which exercise this particular piece of code (you will have to look for which ones)

[the remark this answers is not in the document]
Comment (debnathshoham) [reply pandas/core/frame.py:4283]: @jreback added another relevant test. Please let me know if this is fine.

[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/frame.py:4283]: While the added tests were good. I think what was meant was tests for the `select_dtypes` method specifically.

[the remark this answers is not in the document]
Comment (debnathshoham) [reply pandas/core/frame.py]: @mroeschke added test for `select_dtypes`. Please let me know if this is ok?

File: pandas/tests/window/test_rolling.py [hunk 6]
```diff
@@ -1424,3 +1424,11 @@ def test_rolling_zero_window():
     result = s.rolling(0).min()
     expected = Series([np.nan])
     tm.assert_series_equal(result, expected)
+
+
+def test_rolling_float_dtype(float_dtype):
```
Comment (simonjayhawkins) [inline pandas/tests/window/test_rolling.py:1429]: this worked in 1.1.5, see https://github.com/pandas-dev/pandas/issues/42452#issuecomment-891732335

we may want to consider backporting this fix.

**Comment by simonjayhawkins** [comment]
@debnathshoham does this fix also close #41779

**Comment by debnathshoham** [comment]
hi @simonjayhawkins,
I am getting the below right now (i.e. columns a with float16 is getting dropped)
```
      b     c     d     e     f
0   1.0   1.0   2.0   3.0   4.0
1   7.0   7.0   8.0   9.0  10.0
2  13.0  13.0  14.0  15.0  16.0
3  19.0  19.0  20.0  21.0  22.0
```

Adding float16 as well to the `converted_dtypes` list, would fix that as well.

**Comment by debnathshoham** [comment]
alternatively, i can see that just replacing with "number" below also fixes both (as suggested in the other BUG)
`obj = obj.select_dtypes(include=["integer", "float"], exclude=["timedelta"])`

**Comment by jreback** [comment]
we are not supporting float16 in any way

**Comment by debnathshoham** [comment]
Hi @jreback, then this patch should work fine. Pls let me know if you think of any other changes I should make.

**Review by jreback — APPROVED** [review]
lgtm. @mroeschke

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @debnathshoham

**Pull request merged**